File: Functional.scala
val add2 = (y : Int) => y + 2
val add3 = (y : Int) => y + 3


val compose = (f : Int => Int, g: Int => Int) => (x : Int) => g(f(x))
val add5Composed = compose(add2, add3)

println("Simple Function --- add2(8) = %d".format(add2(8)))
println("Function composition using BinaryOperator<Function> --- add5Composed(5) = %d".format(add5Composed(5)))
println("Function Composition using .compose() --- add3.compose(add2)(5) = %d".format(add3.compose(add2)(5))); //10

val adder = (y : Int) => (x : Int) => x + y
val add2Curry = adder(2)

println("Currying Function --- add2Curry(8) = %d".format(add2Curry(8)))
println("Function Application --- adder(2)(8) = %d".format(adder(2)(8)))

File: Functional.java
import java.util.function.*;

class Functional{

    public static void main(String[] args){
        Function<Integer, Integer> add2 = y -> 2 + y,
                                    add3 = y -> 3 + y;


        BinaryOperator<Function<Integer, Integer>> compose = (f ,g) -> x -> g.apply(f.apply(x));  // h(x) = f(g(x))
        // BinaryOperator<Function<Integer, Integer>> compose = new BinaryOperator<Function<Integer, Integer>>(){
        //     @Override
        //     public Function<Integer, Integer> apply(Function<Integer, Integer> f, Function<Integer, Integer> g){
        //         return new Function<Integer, Integer>(){
        //             @Override
        //             public Integer apply(Integer x){
        //                 return g.apply(f.apply(x));
        //             }
        //         };
        //     }
        // };

        Function<Integer,Integer> add5Composed = compose.apply(add2, add3);

        System.out.println(String.format("Simple Function --- add2.apply(8) = %d", add2.apply(8))); //10
        System.out.println(String.format("Function<T,R> Compositiopn using BinaryOperator --- add5Composed.apply(5)) = %d", add5Composed.apply(5))); //10
        System.out.println(String.format("Function<T,R> Composition using .compose() --- add3.compose(add2).apply(5)) = %d", add3.compose(add2).apply(5)) ); //10


        Function<Integer, Function<Integer, Integer>> makeAdder = x -> y -> x + y;
        //Function<Integer, Function<Integer, Integer>> makeAdder = Functional::adder;

        Function<Integer, Integer> add2Curry = makeAdder.apply(2); //makeAdder 2



        System.out.println(String.format("Currying Function --- add2Curry.apply(8) = %d", add2Curry.apply(8))); // 10
        System.out.println(String.format("Function Application --- makeAdder.apply(2).apply(8) = %d", makeAdder.apply(2).apply(8))); //makeAdder 2 8 = 10

    }

    public static Function<Integer, Integer> adder(Integer x) {
       return y -> x + y;
    }
}
